### DISCUSSION TITLE
Run tests without commit - Buildbot try

### DISCUSSION THREAD

**Discussion by adiroiban** [community/community#25940] [Actions]
<p>Hi,</p>
<p>I am wondering if there is already a way to run a test suite inside a workflow without pushing a commit.</p>
<p>Here is what I got so far… is all open source <a href="https://github.com/chevah/brink/pull/120" rel="nofollow noopener">https://github.com/chevah/brink/pull/120</a> , but is based on the legacy buildbot code designed to run on AIX 5.3, HPUX and Solaris 10 on Sparc.</p>
<p><a href="https://asciinema.org/a/nn1U7FLtbyWm57b0dT7U5sKD0" rel="nofollow noopener"><img src="https://asciinema.org/a/nn1U7FLtbyWm57b0dT7U5sKD0.svg" alt="asciicast" width="690" height="350"></a></p>
<p>I am using buildbot try (<a href="https://docs.buildbot.net/current/manual/cmdline.html#try" rel="nofollow noopener">https://docs.buildbot.net/current/manual/cmdline.html#try</a>) command to help with TDD for multi-OS code.</p>
<p>Most of the time my usage is as follows.<br>
I develop on my local OS (ex Linux). I make sure all the tests pass.<br>
I then push a commit.<br>
I find out that some tests are failing on Windows or MacOS.<br>
I then start using <code>buildbot try</code> to just re-run the selected tests for a patch without a commit<br>
The tests are run on private hosted VM which are not reset at the end of the test run.<br>
So is like a remote command execution.</p>
<hr>
<p>I did a proof of concept tool for GitHub actions using the workflow_dispatch event which has a <code>diff</code> as input.<br>
I create a local diff for uncommitted changed, encode it in base64 and then send it to the workflow.</p>
<p>Then the workflow has the required rules for decoding and applying the diff.</p>
<p>For now is a big hack, as it looks like the current API is not designed for such an usage.<br>
When I trigger the workflow_dispatch via the REST API I don’t get the run ID as part of the response.</p>
<hr>
<p>What do you think?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/adiroiban">@adiroiban</a>,</p>
<blockquote>
<p>When I trigger the workflow_dispatch via the REST API I don’t get the run ID as part of the response.</p>
</blockquote>
<p>Yeah, currently the “<a href="https://docs.github.com/en/rest/reference/actions#create-a-workflow-dispatch-event" rel="nofollow noopener"><strong>Create a workflow dispatch event</strong></a>” API only returns a status code (e.g. <strong>204 No Content</strong>) in the <strong>Response</strong>. Just like the “<a href="https://docs.github.com/en/rest/reference/repos#create-a-repository-dispatch-event" rel="nofollow noopener"><strong>Create a repository dispatch event</strong></a>” API (for <a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#repository_dispatch" rel="nofollow noopener"><code>repository_dispatch</code></a>).</p>
<p>If you want to view the detailed information of the triggered workflow run, you can output the <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener"><strong>github</strong> context</a> in the workflow run.</p>
<pre><code class="lang-yaml">- name: Dump GitHub context
  env:
    GITHUB_CONTEXT: ${{ toJson(github) }}
  run: echo "$GITHUB_CONTEXT"
</code></pre>
<p>Of course, if your projects really need the feature that the “<strong>Create a workflow dispatch event</strong>” API can return the id of the triggered workflow run in the <strong>Response</strong>, I recommend that you can directly report a feature request <a href="https://support.github.com/contact/feedback?contact" rel="nofollow noopener">here</a>. That will allow you to directly interact with the appropriate engineering team, and make it more convenient for the engineering team to collect and categorize your suggestions.</p>

**Comment by adiroiban** [comment]
<p>Thanks for your response.<br>
I have already made a feature request for the run ID.<br>
For now, constantly checking for new jobs works as we don’t have multiple person running the event at the same time for the same branch.</p>
<hr>
<p>I prefer this way of dumping the context, as I get a nicely formatted json.</p>
<pre><code class="lang-yaml">    - name: Show action context
      run: |
        echo &lt;&lt; EOF
        ${{ toJSON(github) }}
        EOF
</code></pre>
<hr>
<p>Do you know why using following conditions doesn’t work</p>
<pre><code class="lang-auto">  linux_unicode_path:
    runs-on: ubuntu-20.04
    if: github.event.inputs.job == '' || github.event.inputs.job == github.job
</code></pre>
<p>But if I use something like this works?</p>
<pre><code class="lang-auto">  linux_unicode_path:
    runs-on: ubuntu-20.04
    if: github.event.inputs.job == '' || github.event.inputs.job == 'linux_unicode_path'
</code></pre>
<p>I have checked the Github context value, and  github.event.inputs.job and github.job are set to the same value.</p>
<p>Thanks</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/adiroiban">@adiroiban</a>,</p>
<blockquote>
<p><strong>Do you know why using following conditions doesn’t work<br>
linux_unicode_path:<br>
runs-on: ubuntu-20.04<br>
if: github.event.inputs.job == ‘’ || github.event.inputs.job == github.job</strong></p>
</blockquote>
<p>There are few points you may need to understand:</p>
<ol>
<li>
<p>The id of current job won’t be assigned to the property ‘<strong>github.job</strong>’ until the job starts up.<br>
When you use the expression ‘<strong>github.job</strong>’ on the job level <strong>if</strong> conditional, it will always return ‘<strong>null</strong>’ since the current job has not started up.<br>
So, you can only use the expression ‘<strong>github.job</strong>’ in the steps (include the step level <strong>if</strong> conditional) after the current job starts up.</p>
</li>
<li>
<p>For the expression ‘<strong>github.event.inputs.job</strong>’, if the current workflow run is not triggered on the ‘<strong>workflow_dispatch</strong>’ event, the <a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener"><strong>github</strong> context</a> will not contain the object ‘<strong>github.event.inputs</strong>’.<br>
So, when the workflow is triggered on other events,  the expression ‘<strong>github.event.inputs.job</strong>’ or ‘<strong>github.event.inputs</strong>’ may always return ‘<strong>null</strong>’, and the if statement (<strong>github.event.inputs.job == ‘’</strong>) will always return ‘<strong>true</strong>’.</p>
</li>
</ol>
<p>To make sure the job ‘<strong>linux_unicode_path</strong>’ in your workflow only run when the current workflow run is triggered on the ‘<strong>workflow_dispatch</strong>’ event, and can meet the conditional (<strong>github.event.inputs.job == ‘’</strong>) or (<strong>github.event.inputs.job == linux_unicode_path</strong>), you can change the job level <strong>if</strong> conditional to like below:</p>
<pre><code class="lang-yaml">linux_unicode_path:
    runs-on: ubuntu-20.04
    if: ${{ github.event_name == 'workflow_dispatch' &amp;&amp; (github.event.inputs.job == '' || github.event.inputs.job == 'linux_unicode_path') }}
</code></pre>
